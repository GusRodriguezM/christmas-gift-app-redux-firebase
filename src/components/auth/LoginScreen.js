import React, { useMemo, useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';

import { startGoogleSignIn, startLoginWithEmailPassword } from '../../store/slices/auth';
import { removeErrorMessage, setErrorMessage } from '../../store/slices/ui';
import { useForm } from '../../hooks';

import { Button } from '../styles/shared/Button.styled';
import Container from '../styles/auth/Container.styled';
import Input from '../styles/elements/Input.styled';
import Toast from '../styles/Toast/Toast';

export const LoginScreen = () => {

    const { status } = useSelector( state => state.auth );
    const { msgError } = useSelector( state => state.ui );
    const dispatch = useDispatch();

    const [formValues, handleInputChange] = useForm({
        email: '',
        password: ''
    });

    const { email, password } = formValues;

    const isAuthenticating = useMemo(() => status === 'checking', [status]);

    useEffect(() => {
        if(msgError.length > 0){
            Toast.fire({
                icon: 'error',
                title: `${msgError}`
            });
        }
    }, [msgError]);

    const isFormValid = () => {
        if(validator.isEmpty(email) && validator.isEmpty(password)){
            dispatch( setErrorMessage('Email and password are required') );
            return false;
        }else if(validator.isEmpty(email)){
            dispatch( setErrorMessage('The email is required') );
            return false;
        }else if(!validator.isEmail(email)){
            dispatch( setErrorMessage('The email is not valid') );
            return false;
        }else if(validator.isEmpty(password)){
            dispatch( setErrorMessage('The password is required') );
            return false;
        }

        dispatch( removeErrorMessage() );
        return true;
    }

    const handleLogin = (e) => {
        e.preventDefault();
        
        if(isFormValid())
            dispatch( startLoginWithEmailPassword({email, password}) );
    }

    const handleGoogleSignIn = () => {
        dispatch( startGoogleSignIn() );
    }

    return (
        <Container>

            <h1>Login</h1>

            <Container.AuthForm onSubmit={handleLogin}>

                <Input
                    id='name'
                    type='text'
                    placeholder='Email'
                    name='email'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />

                <Input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    autoComplete='off'
                    value={password}
                    onChange={handleInputChange}
                />

                <Button 
                    type='submit'
                    disabled={isAuthenticating}
                    inactive={isAuthenticating}
                >
                    <span>Login</span>
                    <i className="fa-solid fa-user-lock"></i>
                </Button>

                <Button 
                    onClick={handleGoogleSignIn}
                    disabled={isAuthenticating}
                    inactive={isAuthenticating}
                >
                    <span>Google Login</span>
                    <i className="fa-brands fa-google"></i>
                </Button>

            </Container.AuthForm>

            <Link
                aria-disabled={isAuthenticating}
                to='/auth/register'
            >
                Create a new account
            </Link>
        </Container>
    )
}