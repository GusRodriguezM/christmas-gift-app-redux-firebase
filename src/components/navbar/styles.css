.navbar {
    height: 60px;
    width: 100%;
    display: flex;
    align-items: center;
    position: relative;
    padding: 0.5rem 0rem;
    background-color: #e8ceac;
    color: #000000;
    box-shadow: 0 2px 2px 2px rgba(9, 9, 9, 0.23);
}

.navbar__brand-name {
    color: #000000;
    margin-left: 1rem;
    text-decoration: none;
}

.navbar__hamburger {
    border: 0;
    height: 40px;
    width: 40px;
    padding: 0.5rem;
    border-radius: 50%;
    background-color: #283b8b;
    cursor: pointer;
    transition: background-color 0.2s ease-in-out;
    position: absolute;
    top: 50%;
    right: 25%;
    transform: translateY(-50%);
    display: none;
}

.navbar__hamburger:hover {
    background-color: #2642af;
}

.navbar__navigation-menu {
    margin-left: auto;
}

.navbar__navigation-menu ul {
    display: flex;
    padding: 0;
}

.navbar__navigation-menu li {
    list-style-type: none;
    margin: 0 1rem;
}

.navbar__navigation-menu li a {
    text-decoration: none;
    display: block;
    width: 100%;
}