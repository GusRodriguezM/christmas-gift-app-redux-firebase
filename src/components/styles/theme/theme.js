const theme = {
    colors: {
        black: '#000000',
        loginButton: '#4285f4',
        disabledButton: '#999595',
        red: '#C30F16',
        white: '#FFFFFF',
        list: '#c8e4e4',
        navbar: '#e8ceac',
        boxShadow: 'rgba(9, 9, 9, 0.23)',
    }
}

export default theme;