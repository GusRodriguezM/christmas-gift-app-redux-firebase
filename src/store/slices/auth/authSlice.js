import { createSlice } from '@reduxjs/toolkit';

export const authSlice = createSlice({
  name: 'auth',
  initialState: {
    status: 'checking', //options: checking, non-authenticated, authenticated
    uid: null,
    email: null,
    userName: null,
    photoURL: null,
    errorMessage: null,
    logged: false
  },
  reducers: {
    login: (state, { payload }) => {
        state.status = 'authenticated';
        state.uid = payload.uid;
        state.email = payload.email;
        state.userName = payload.displayName;
        state.photoURL = payload.photoURL;
        state.errorMessage = payload.errorMessage;
    },
    logout: (state, { payload }) => {
        state.status = 'non-authenticated';
        state.uid = null;
        state.email = null;
        state.userName = null;
        state.photoURL = null;
        state.errorMessage = payload.errorMessage;
    },
    checkingUserCredentials: (state, action) => {
      state.status = 'checking';
    },
  },
});

export const { login, logout, checkingUserCredentials } = authSlice.actions;