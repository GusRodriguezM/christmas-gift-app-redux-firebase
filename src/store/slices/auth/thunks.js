import { signInWithGoogle } from "../../../firebase/providers";
import { checkingUserCredentials, login, logout } from "./"

export const checkingAuthentication = (email, password) => {
    return async( dispatch ) => {
        dispatch( checkingUserCredentials() );
    }
}

export const startGoogleSignIn = () => {
    return async( dispatch ) => {
        dispatch( checkingUserCredentials() );
        const result = await signInWithGoogle();
        
        if(!result.ok)
            return dispatch( logout( result.errorMessage ) );

        dispatch( login(result) );
    }
}