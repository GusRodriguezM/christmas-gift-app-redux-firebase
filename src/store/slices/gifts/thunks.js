import { collection, deleteDoc, doc, setDoc, writeBatch } from 'firebase/firestore/lite';
import { FirebaseDB } from '../../../firebase/config';
import { loadGifts } from '../../../helpers';
import { addGift, setSavingGift, setGifts, editGift, deleteGiftById, cleanList } from './';

export const startAddingNewGift = (newGift) => {
    return async(dispatch, getState) => {

        dispatch( setSavingGift() );
        
        const { uid } = getState().auth;

        const newDoc = doc( collection( FirebaseDB, `${uid}/list/gifts` ) );
        await setDoc( newDoc, newGift );

        newGift.id = newDoc.id;

        dispatch( addGift(newGift) );

    }
}

export const startLoadingGifts = () => {
    return async(dispatch, getState) => {

        const { uid } = getState().auth;
        if(!uid) throw new Error('The user id does not exist');

        const gifts = await loadGifts(uid);

        dispatch( setGifts(gifts) );
    }
}

export const startSavingGift = (giftToEdit) => {
    return async(dispatch, getState) => {
        dispatch( setSavingGift() );

        const { uid } = getState().auth;
        const { activeGift } = getState().gifts;

        const docRef = doc(FirebaseDB, `${uid}/list/gifts/${activeGift.id}`);
        await setDoc(docRef, giftToEdit, { merge: true });

        giftToEdit.id = activeGift.id;
        dispatch( editGift(giftToEdit) );
    }
}

export const startDeletingGift = (id) => {
    return async(dispatch, getState) => {

        const { uid } = getState().auth;

        const docRef = doc(FirebaseDB, `${uid}/list/gifts/${id}`);
        await deleteDoc(docRef);

        dispatch( deleteGiftById(id) );
        
    }
}

export const startDeletingGifts = () => {
    return async(dispatch, getState) => {
        
        //Getting the user id and the gifts array from the store
        const { uid } = getState().auth;
        const { gifts } = getState().gifts;

        //We use batch writes. Basically this instruction allows to execute a set of different operations
        const batch = writeBatch(FirebaseDB);

        //Loop over the array of gifts to delete. We send the reference to the doc
        gifts.forEach(gift => {
            batch.delete( doc(FirebaseDB, `${uid}/list/gifts/${gift.id}`) );
        });

        //Finally, we commit the batch
        await batch.commit();

        //We clean the array in the store
        dispatch( cleanList() );

    }
}