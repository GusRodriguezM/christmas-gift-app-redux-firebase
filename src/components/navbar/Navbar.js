import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../store/slices/auth';
import { BrandName, NavBar, NavContent, NavInfo } from '../styles/navbar/Navbar.styled';
import { Button } from '../styles/shared/Button.styled';
import { Span } from '../styles/shared/Span.styled';

export const Navbar = () => {

    const { status, userName } = useSelector( state => state.auth );
    const dispatch = useDispatch();
    const [isNavExpanded, setIsNavExpanded] = useState(false);

    const handleLogout = (e) => {
        e.preventDefault();
        dispatch( startLogout() );
    }

    return (
        <NavBar>
            <BrandName to='/' >
                Christmas Gift App
            </BrandName>

            {/* <button className='navbar__hamburger' onClick={() => setIsNavExpanded(!isNavExpanded)}>
                <i className="fa-solid fa-bars"></i>
            </button> */}

            <NavContent>
                {
                    status === 'authenticated' && (
                        <NavInfo>
                            <Span>
                                { userName }
                            </Span>

                            <Button
                                onClick={handleLogout}
                            >
                                <span>Logout</span>
                                <i className="fa-solid fa-right-from-bracket"></i>
                            </Button>
                        </NavInfo>
                    )
                }
            </NavContent>
        </NavBar>
    )
}