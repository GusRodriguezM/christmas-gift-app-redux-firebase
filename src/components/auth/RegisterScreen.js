import React from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import { login } from '../../store/slices/auth/authSlice';
import { useForm } from '../hooks/useForm';
import Container from '../styles/auth/Container.styled';
import Input from '../styles/elements/Input.styled';
import { Button } from '../styles/shared/Button.styled';

export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const navigate = useNavigate();

    const [formValues, handleInputChange] = useForm({
        name: '',
        email: '',
        password: ''
    });

    const { name, email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        
        dispatch( login(email) );
        navigate('/', {
            replace: true
        });
    }

    const handleNavigate = () => {
        navigate('/auth/login', {
            replace: true
        });
    }

    return (
        <Container>

            <h1 className='login__title'>Sign in</h1>

            <Container.AuthForm onSubmit={handleLogin} className='login__form'>

                <Input
                    type='text'
                    placeholder='Name'
                    name='name'
                    className='login__form--input'
                    autoComplete='off'
                    value={name}
                    onChange={handleInputChange}
                />

                <Input 
                    type='text'
                    placeholder='Email'
                    name='email'
                    className='login__form--input'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />

                <Input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    className='login__form--input'
                    autoComplete='off'
                    value={password}
                    onChange={handleInputChange}
                />

                <Button type='submit'>
                    <span>Login</span>
                </Button>

            </Container.AuthForm>

            <Container.Span>I already have an account</Container.Span>
            <Button
                onClick={handleNavigate}
            >
                Login
            </Button>
        </Container>
    )
}
