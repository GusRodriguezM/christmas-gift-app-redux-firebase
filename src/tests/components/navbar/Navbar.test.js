import { configureStore } from "@reduxjs/toolkit";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { Navbar } from "../../../components/navbar/Navbar";
import theme from "../../../components/styles/theme/theme";
import { authSlice } from "../../../store/slices/auth/authSlice";
import { startLogout } from "../../../store/slices/auth/thunks";
import { authenticatedState } from "../../fixtures/authFixtures";

//Mocking the thunk to begin the process to the logout
const mockStartLogout = jest.fn();
jest.mock('../../../store/slices/auth/thunks', () => ({
    startLogout: () => mockStartLogout
}));

describe('Tests in the <Navbar />', () => {

    //Mocking the store with an auth state
    const store = configureStore({
        reducer: {
            auth: authSlice.reducer
        },
        preloadedState: {
            auth: authenticatedState
        }
    });

    //Cleaning the mocks
    beforeEach( () => jest.clearAllMocks() );

    test('Should render the component correctly and match with the snapshot', () => {

        //Rendering the component
        const { container } = render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <Navbar />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );

        //Expecting that the result of render the component matches with the snapshot
        expect( container ).toMatchSnapshot();

        //Expecting that the name of the user appears in the navbar
        expect( screen.getByText(`: ${store.getState().auth.userName}`) ).toBeTruthy();

    });

    test('Logout button should call the startLogout action', async() => {

        //Rendering the component
        render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <Navbar />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );

        //Getting the reference of the button
        const logoutBtn = screen.getByRole('button', { name: 'Logout' });

        //Firing the event to make the logout
        fireEvent.click( logoutBtn );

        //We expect that the startLogout have been called after clicking the logout button
        expect( mockStartLogout ).toHaveBeenCalled();

    });

});