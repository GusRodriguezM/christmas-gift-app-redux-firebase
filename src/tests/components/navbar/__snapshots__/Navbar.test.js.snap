// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in the <Navbar /> Should render the component correctly and match with the snapshot 1`] = `
<div>
  <nav
    class="sc-bczRLJ gslatN"
  >
    <a
      class="sc-gsnTZi dYciIz"
      href="/"
    >
      Christmas Gift App
    </a>
    <div
      class="sc-dkzDqf jhiqnR"
    >
      <span
        class="sc-jSMfEi iRIMqE"
      >
        <i
          class="fa-solid fa-user-astronaut"
        />
        : Jon Doe
      </span>
      <button
        class="sc-hKMtZM gatnXf"
      >
        <span>
          Logout
        </span>
        <i
          class="fa-solid fa-right-from-bracket"
        />
      </button>
    </div>
  </nav>
</div>
`;
